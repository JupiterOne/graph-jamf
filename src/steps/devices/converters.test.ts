import {
  createComputerEntity,
  createMobileDeviceEntity,
  getDeploymentStatus,
} from './converters';
import {
  createMockMobileDevice,
  createMockComputer,
  createMockComputerDetail,
  EXTENSION_ATTRIBUTES,
  createMockMobileDeviceDetail,
} from '../../../test/mocks';

describe('#createMobileDeviceEntity', () => {
  test('should convert data', () => {
    expect(
      createMobileDeviceEntity(
        createMockMobileDevice(),
        createMockMobileDeviceDetail(),
        false,
      ),
    ).toMatchSnapshot();
  });
});

describe('#createComputerEntity', () => {
  test('should convert data', () => {
    const data = createComputerEntity({
      device: createMockComputer(),
      macOsConfigurationDetailByIdMap: new Map(),
      detailData: createMockComputerDetail(),
      previouslyDiscoveredDevice: false,
    });
    expect(data).toMatchSnapshot();
  });
});

describe('#getDeploymentStatus', () => {
  test('should get the value of the deployment status hash', () => {
    expect(getDeploymentStatus(EXTENSION_ATTRIBUTES)).toEqual(['active']);
  });
  test('should handle null', () => {
    expect(getDeploymentStatus(null as any)).toEqual([]);
  });
  test('should handle undefined', () => {
    expect(getDeploymentStatus(undefined as any)).toEqual([]);
  });
  test('should handle incorrect input', () => {
    expect(getDeploymentStatus(2 as any)).toEqual([]);
  });
});
