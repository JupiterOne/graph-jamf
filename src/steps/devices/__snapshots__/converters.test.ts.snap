// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createComputerEntity should convert data 1`] = `
Object {
  "_class": Array [
    "Host",
    "Device",
  ],
  "_key": "ABC123DEF456",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "building": "",
        "department": "",
        "id": 1,
        "mac_address": "12:34:67:8A:9B:1C",
        "managed": true,
        "model": "MacBookPro13,3",
        "name": "Johnâ€™s MacBook",
        "report_date_epoch": 1560787256384,
        "report_date_utc": "2019-06-17T16:00:56.384+0000",
        "serial_number": "ABC123DEF456",
        "udid": "DF7C48EF-D17E-4A5E-B192-8C758870F624",
        "username": "john",
      },
    },
    Object {
      "name": "detail",
      "rawData": Object {
        "certificates": Array [
          Object {
            "common_name": "com.apple.kerberos.kdc",
            "expires_epoch": "",
            "expires_utc": "",
            "identity": true,
            "name": "",
          },
          Object {
            "common_name": "Apple Worldwide Developer Relations Certification Authority",
            "expires_epoch": 1675806527000,
            "expires_utc": "2023-02-07T21:48:47.000+0000",
            "identity": false,
            "name": "",
          },
        ],
        "configuration_profiles": Array [
          Object {
            "id": -2,
            "is_removable": false,
            "name": "",
            "uuid": "",
          },
          Object {
            "id": -1,
            "is_removable": false,
            "name": "",
            "uuid": "",
          },
        ],
        "extension_attributes": Array [
          Object {
            "id": 1,
            "name": "123",
            "value": Array [
              "foo123",
              "bar123",
            ],
          },
          Object {
            "id": 2,
            "name": "456",
            "value": Array [
              "foo456",
              "bar456",
            ],
          },
          Object {
            "id": 3,
            "name": "Deployment status",
            "value": Array [
              "active",
            ],
          },
        ],
        "general": Object {
          "alt_mac_address": "12:34:56:78:90:12",
          "alt_network_adapter_type": "Ethernet",
          "asset_tag": "",
          "barcode_1": "",
          "barcode_2": "",
          "distribution_point": "",
          "id": 1,
          "initial_entry_date": "2020-06-03",
          "initial_entry_date_epoch": 1591217314697,
          "initial_entry_date_utc": "2020-06-03T20:48:34.697+0000",
          "ip_address": "12.34.567.890",
          "itunes_store_account_is_active": true,
          "jamf_version": "10.23.0-t1595614145",
          "last_cloud_backup_date_epoch": 0,
          "last_cloud_backup_date_utc": "",
          "last_contact_time": "2020-08-09 13:24:00",
          "last_contact_time_epoch": 1596979440636,
          "last_contact_time_utc": "2020-08-09T13:24:00.636+0000",
          "last_enrolled_date_epoch": 1591217314697,
          "last_enrolled_date_utc": "2020-06-03T20:48:34.697+0000",
          "last_reported_ip": "192.168.1.139",
          "mac_address": "12:34:67:8A:9B:1C",
          "management_status": Object {
            "enrolled_via_dep": false,
            "user_approved_enrollment": false,
            "user_approved_mdm": false,
          },
          "mdm_capable": true,
          "mdm_capable_users": Object {
            "mdm_capable_user": "john.doe",
          },
          "mdm_profile_expiration_epoch": 1748983791000,
          "mdm_profile_expiration_utc": "2025-06-03T20:49:51.000+0000",
          "name": "John's MacBook",
          "netboot_server": "",
          "network_adapter_type": "IEEE80211",
          "platform": "Mac",
          "remote_management": Object {
            "managed": true,
            "management_password_sha256": "abc123",
            "management_username": "macOS",
          },
          "report_date": "2020-08-03 12:24:29",
          "report_date_epoch": 1596457469968,
          "report_date_utc": "2020-08-03T12:24:29.968+0000",
          "serial_number": "ABC123DEF456",
          "site": Object {
            "id": 1,
            "name": "Test site",
          },
          "supervised": false,
          "sus": "",
          "udid": "DF7C48EF-D17E-4A5E-B192-8C758870F624",
        },
        "groups_accounts": Object {
          "computer_group_memberships": Array [
            "All Managed Clients",
          ],
          "local_accounts": Array [
            Object {
              "administrator": true,
              "filevault_enabled": true,
              "home": "/Users/john",
              "home_size": "-1MB",
              "home_size_mb": -1,
              "name": "john.doe",
              "realname": "John",
              "uid": "501",
            },
            Object {
              "administrator": true,
              "filevault_enabled": true,
              "home": "/Users/macOS",
              "home_size": "-1MB",
              "home_size_mb": -1,
              "name": "macOS",
              "realname": "macOS",
              "uid": "502",
            },
          ],
          "user_inventories": Object {
            "disable_automatic_login": true,
            "user": Object {
              "password_history_depth": "",
              "password_max_age": "",
              "password_min_complex_characters": "",
              "password_min_length": "4",
              "password_require_alphanumeric": "false",
              "username": "macOS",
            },
          },
        },
        "hardware": Object {
          "active_directory_status": "Not Bound",
          "available_ram_slots": 0,
          "battery_capacity": 93,
          "ble_capable": true,
          "boot_rom": "204.0.0.0.0",
          "bus_speed": 0,
          "bus_speed_mhz": 0,
          "cache_size": 8192,
          "cache_size_kb": 8192,
          "disk_encryption_configuration": "",
          "filevault2_users": Array [
            "macOS",
            "john.doe",
          ],
          "gatekeeper_status": "App Store and identified developers",
          "institutional_recovery_key": "Not Present",
          "make": "Apple",
          "mapped_printers": Array [
            Object {
              "location": "",
              "name": "Brother ABC",
              "type": "Brother ABC",
              "uri": "dnssd://Brother%20ABC._pdl-datastream._tcp.local./?bidi",
            },
          ],
          "master_password_set": false,
          "model": "15-inch Retina MacBook Pro with TouchID (Mid 2017)",
          "model_identifier": "MacBookPro14,3",
          "nic_speed": "n/a",
          "number_cores": 4,
          "number_processors": 1,
          "optical_drive": "",
          "os_build": "19D76",
          "os_name": "Mac OS X",
          "os_version": "10.15.3",
          "processor_architecture": "x86_64",
          "processor_speed": 3100,
          "processor_speed_mhz": 3100,
          "processor_type": "Quad-Core Intel Core i7",
          "service_pack": "",
          "sip_status": "Enabled",
          "smc_version": "2.45f1",
          "storage": Array [
            Object {
              "connection_type": "NO",
              "disk": "disk0",
              "drive_capacity_mb": 1000555,
              "model": "APPLE SSD SM1024L",
              "partitions": Array [
                Object {
                  "available_mb": 207295,
                  "boot_drive_available_mb": 207295,
                  "filevault2_percent": 100,
                  "filevault2_status": "Encrypted",
                  "filevault_percent": 100,
                  "filevault_status": "Encrypted",
                  "lvUUID": "",
                  "lvgUUID": "",
                  "name": "Macintosh HD (Boot Partition)",
                  "partition_capacity_mb": 1000240,
                  "percentage_full": 5,
                  "pvUUID": "",
                  "size": 1000240,
                  "type": "boot",
                },
                Object {
                  "available_mb": 207295,
                  "filevault2_percent": 100,
                  "filevault2_status": "Encrypted",
                  "filevault_percent": 100,
                  "filevault_status": "Encrypted",
                  "name": "Data",
                  "partition_capacity_mb": 1000240,
                  "percentage_full": 79,
                  "size": 1000240,
                  "type": "other",
                },
              ],
              "revision": "CXS5EA0Q",
              "serial_number": "123ABC",
              "size": 1000555,
              "smart_status": "Verified",
            },
          ],
          "total_ram": 16384,
          "total_ram_mb": 16384,
          "xprotect_version": "2127",
        },
        "iphones": Array [],
        "location": Object {
          "building": "",
          "department": "",
          "email_address": "JOHN.doe@test.com",
          "phone": "111",
          "phone_number": "111",
          "position": "CISO",
          "real_name": "John Doe",
          "realname": "John Doe",
          "room": "",
          "username": "john.doe",
        },
        "peripherals": Array [],
        "purchasing": Object {
          "applecare_id": "",
          "attachments": Array [],
          "is_leased": false,
          "is_purchased": true,
          "lease_expires": "",
          "lease_expires_epoch": 0,
          "lease_expires_utc": "",
          "life_expectancy": 0,
          "os_applecare_id": "",
          "os_maintenance_expires": "",
          "po_date": "",
          "po_date_epoch": 0,
          "po_date_utc": "",
          "po_number": "",
          "purchase_price": "",
          "purchasing_account": "",
          "purchasing_contact": "",
          "vendor": "",
          "warranty_expires": "",
          "warranty_expires_epoch": 0,
          "warranty_expires_utc": "",
        },
        "security": Object {
          "activation_lock": false,
          "external_boot_level": "not supported",
          "firewall_enabled": false,
          "secure_boot_level": "not supported",
        },
        "software": Object {
          "applications": Array [
            Object {
              "bundle_id": "com.github.atom",
              "name": "Atom.app",
              "path": "/Applications/Atom.app",
              "version": "1.43.0",
            },
          ],
          "available_software_updates": Array [],
          "available_updates": Object {},
          "cached_by_casper": Array [],
          "fonts": Array [],
          "installed_by_casper": Array [],
          "installed_by_installer_swu": Array [
            "com.apple.pkg.Core",
          ],
          "licensed_software": Array [],
          "plugins": Array [],
          "running_services": Array [
            "com.apple.accessoryd",
          ],
          "unix_executables": Array [],
        },
      },
    },
  ],
  "_type": "user_endpoint",
  "altMacAddress": "12:34:56:78:90:12",
  "altNetworkAdapterType": "Ethernet",
  "assetTag": "",
  "barcode1": "",
  "barcode2": "",
  "building": "",
  "category": "endpoint",
  "createdOn": 1591217314697,
  "department": "",
  "deviceId": "DF7C48EF-D17E-4A5E-B192-8C758870F624",
  "deviceType": "laptop",
  "displayName": "Johnâ€™s MacBook",
  "distributionPoint": "",
  "email": "john.doe@test.com",
  "encrypted": true,
  "enrolledOn": 1591217314697,
  "extensionAttribute.Deployment status": Array [
    "active",
  ],
  "firewallEnabled": false,
  "gatekeeperEnabled": true,
  "gatekeeperStatus": "App Store and identified developers",
  "hostname": null,
  "id": "1",
  "ipAddress": "12.34.567.890",
  "itunesStoreAccountIsActive": true,
  "jamfVersion": "10.23.0-t1595614145",
  "lastCloudBackupDateEpoch": 0,
  "lastCloudBackupDateUtc": "",
  "lastReportedIp": "192.168.1.139",
  "lastReportedOn": 1560787256384,
  "lastSeenOn": 1596979440636,
  "macAddress": "12:34:67:8a:9b:1c",
  "make": "Apple",
  "managed": true,
  "mdmCapable": true,
  "mdmProfileExpirationEpoch": 1748983791000,
  "mdmProfileExpirationUtc": "2025-06-03T20:49:51.000+0000",
  "model": "MacBookPro13,3",
  "name": "John's MacBook",
  "netbootServer": "",
  "networkAdapterType": "IEEE80211",
  "osBuild": "19D76",
  "osName": "macOS",
  "osVersion": "10.15.3",
  "platform": "darwin",
  "realName": "John Doe",
  "serial": "ABC123DEF456",
  "serialNumber": "ABC123DEF456",
  "supervised": false,
  "sus": "",
  "systemIntegrityProtectionEnabled": true,
  "udid": "DF7C48EF-D17E-4A5E-B192-8C758870F624",
  "username": "john",
}
`;

exports[`#createMobileDeviceEntity should convert data 1`] = `
Object {
  "_class": Array [
    "Device",
  ],
  "_key": "CA44C68660A3",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "device_name": "Update 1-1",
        "id": 16,
        "managed": true,
        "model": "iPhone 5S (GSM)",
        "model_display": "iPhone 5S (GSM)",
        "model_identifier": "iPhone6,1",
        "name": "Update 1-1",
        "phone_number": "612-356-4364",
        "serial_number": "CA44C68660A3",
        "supervised": false,
        "udid": "ca44c66860a311e490b812df261f2c7e",
        "username": undefined,
        "wifi_mac_address": "0C:3E:9F:49:99:67",
      },
    },
    Object {
      "name": "detail",
      "rawData": Object {
        "applications": Array [
          Object {
            "application": Object {
              "application_name": "Self Service Mobile",
              "application_version": "10.1.1",
              "identifier": "com.jamfsoftware.selfservice",
            },
            "size": 1,
          },
        ],
        "certificates": Array [
          Object {
            "certificate": Object {
              "common_name": "JSS Built-in Certificate Authority",
              "identity": true,
            },
            "size": 1,
          },
        ],
        "configuration_profiles": Array [
          Object {
            "configuration_profile": Object {
              "display_name": "WiFi",
              "identifier": "279889AF-1564-4480-A61D-860DF5EFDF26",
              "uuid": "279889AF-1564-4480-A61D-860DF5EFDF26",
              "version": 1,
            },
            "size": 1,
          },
        ],
        "extension_attributes": Array [
          Object {
            "id": 1,
            "name": "123",
            "value": Array [
              "foo123",
              "bar123",
            ],
          },
          Object {
            "id": 2,
            "name": "456",
            "value": Array [
              "foo456",
              "bar456",
            ],
          },
          Object {
            "id": 3,
            "name": "Deployment status",
            "value": Array [
              "active",
            ],
          },
        ],
        "general": Object {
          "asset_tag": "string",
          "available": 11487,
          "available_mb": 11487,
          "battery_level": 95,
          "ble_capable": true,
          "bluetooth_mac_address": "E0:AC:CB:97:36:G6",
          "capacity": 12159,
          "capacity_mb": 12159,
          "cloud_backup_enabled": true,
          "device_locator_service_enabled": true,
          "device_name": "Tinas iPad",
          "device_ownership_level": "Institutional",
          "display_name": "Tinas iPad",
          "do_not_disturb_enabled": true,
          "exchange_activesync_device_identifier": "TUCLLFJHPL779ACL9DCJQFN39F",
          "id": 1,
          "initial_entry_date_epoch": 1499470624555,
          "initial_entry_date_utc": "2017-07-07T18:37:04.555-0500",
          "ip_address": "192.0.0.1",
          "itunes_store_account_is_active": true,
          "last_backup_time_epoch": 1499470624555,
          "last_backup_time_utc": "2017-07-07T18:37:04.555-0500",
          "last_cloud_backup_date_epoch": 1499470624555,
          "last_cloud_backup_date_utc": "2017-07-07T18:37:04.555-0500",
          "last_enrollment_epoch": 1499470624555,
          "last_enrollment_utc": "2017-07-07T18:37:04.555-0500",
          "last_inventory_update": "2017-07-07 18:37:04",
          "last_inventory_update_epoch": 1499470624555,
          "last_inventory_update_utc": "2017-07-07T18:37:04.555-0500",
          "location_services_enabled": true,
          "managed": true,
          "model": "iPhone 6S",
          "model_display": "iPhone 6S",
          "model_identifier": "iPhone8,1",
          "model_number": "MKRY2LL",
          "modem_firmware": "2.61.00",
          "name": "Tinas iPad",
          "os_build": "14F89",
          "os_type": "iOS",
          "os_version": "10.3.2",
          "percentage_used": 5,
          "phone_number": "123-555-6789",
          "serial_number": "C02Q7KHTGFWF",
          "shared": "string",
          "supervised": true,
          "tethered": "string",
          "udid": "270aae10800b6e61a2ee2bbc285eb967050b5984",
          "wifi_mac_address": "E0:AC:CB:97:36:G4",
        },
        "location": Object {
          "building": "New York Office",
          "department": "Sales Staff",
          "email_address": "jbetty@company.com",
          "phone": "123-555-6789",
          "phone_number": "123-555-6789",
          "position": "Systems Engineer",
          "real_name": "Betty Jackson",
          "realname": "Betty Jackson",
          "room": 1159,
          "username": "JBetty",
        },
        "mobile_device_groups": Array [
          Object {
            "mobile_device_group": Object {
              "id": 1,
              "name": "string",
            },
            "size": 1,
          },
        ],
        "network": Object {
          "carrier_settings_version": 20,
          "cellular_technology": "Unknown",
          "current_carrier_network": "string",
          "current_mobile_country_code": 311,
          "current_mobile_network_code": 480,
          "data_roaming_enabled": true,
          "home_carrier_network": "Verizon",
          "home_mobile_country_code": 311,
          "home_mobile_network_code": 480,
          "iccid": "8914 8000 0010 0254 6259",
          "imei": "35 200706 056227 1",
          "phone_number": 5555555555,
          "voice_roaming_enabled": "Yes",
        },
        "provisioning_profiles": Array [
          Object {
            "mobile_device_provisioning_profile": Object {
              "display_name": "In-House App Provisioning Profile",
              "expiration_date": "2018-01-29",
              "expiration_date_epoch": 0,
              "expiration_date_utc": "2018-01-29T14:18:02.000-0600",
              "uuid": "279889AF-1564-4480-A61D-860DF5EFDF26",
            },
            "size": 1,
          },
        ],
        "purchasing": Object {
          "applecare_id": "string",
          "is_leased": true,
          "is_purchased": true,
          "lease_expires": "string",
          "lease_expires_epoch": 0,
          "lease_expires_utc": "string",
          "life_expectancy": 0,
          "po_date": "string",
          "po_date_epoch": 0,
          "po_date_utc": "string",
          "po_number": "string",
          "purchase_price": "string",
          "purchasing_account": "string",
          "purchasing_contact": "string",
          "vendor": "string",
          "warranty_expires": "string",
          "warranty_expires_epoch": 0,
          "warranty_expires_utc": "string",
        },
        "security": Object {
          "activation_lock_enabled": true,
          "block_level_encryption_capable": true,
          "data_protection": true,
          "file_level_encryption_capable": true,
          "hardware_encryption": "string",
          "jailbreak_detected": "Normal",
          "lost_location_altitude": 242.88067627,
          "lost_location_course": -1,
          "lost_location_epoch": 1517598680459,
          "lost_location_horizontal_accuracy": 65,
          "lost_location_latitude": 44.81436821,
          "lost_location_longitude": -91.502672,
          "lost_location_speed": -1,
          "lost_location_utc": "2018-02-02T13:11:20.459-0600",
          "lost_location_vertical_accuracy": 10,
          "lost_mode_enable_issued_epoch": 1517598768512,
          "lost_mode_enable_issued_utc": "2018-02-02T13:12:48.512-0600",
          "lost_mode_enabled": true,
          "lost_mode_enforced": true,
          "lost_mode_footnote": "string",
          "lost_mode_message": "Please return to Company",
          "lost_mode_phone": "555-555-5555",
          "passcode_compliant": true,
          "passcode_compliant_with_profile": true,
          "passcode_lock_grace_period_enforced": "Not Available",
          "passcode_present": true,
        },
      },
    },
  ],
  "_type": "mobile_device",
  "activationLockEnabled": true,
  "blockLevelEncryption": true,
  "capacity": 12159,
  "category": "mobile",
  "cloudBackupEnabled": true,
  "createdOn": 1499470624555,
  "dataProtectionEnabled": true,
  "deviceId": "ca44c66860a311e490b812df261f2c7e",
  "deviceName": "Update 1-1",
  "displayName": "Unknown User's iPhone 5S (GSM)",
  "email": "jbetty@company.com",
  "extensionAttribute.Deployment status": Array [
    "active",
  ],
  "fileLevelEncryption": true,
  "hasPasscode": true,
  "id": "ca44c66860a311e490b812df261f2c7e",
  "jailbroken": "Normal",
  "lastBackupOn": 1499470624555,
  "lastSeenOn": 1499470624555,
  "locatorServiceEnabled": true,
  "macAddress": "0c:3e:9f:49:99:67",
  "make": null,
  "managed": true,
  "model": "iPhone 5S (GSM)",
  "modelDisplay": "iPhone 5S (GSM)",
  "modelIdentifier": "iPhone6,1",
  "name": "Update 1-1",
  "osType": "iOS",
  "osVersion": "10.3.2",
  "passcodeCompliant": true,
  "passcodeLockGracePeriodEnforced": "Not Available",
  "phoneNumber": "612-356-4364",
  "profileCompliant": true,
  "serial": "CA44C68660A3",
  "serialNumber": "CA44C68660A3",
  "supervised": false,
  "udid": "ca44c66860a311e490b812df261f2c7e",
  "username": undefined,
  "wifiMacAddress": "0C:3E:9F:49:99:67",
}
`;
